[FEAT] adds sandboxed_path

SGSSGene opened the pull request:
This adds the sandboxed_path class which is a safer
std::filesystem::path variant.
It is going to be the base for a refactored tmp_directory class.

simonsasse commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,334 @@
+ *  - sandbox directory is unmutable during the life cycle of a sandboxed_path
+ *
+ * Caveat:
+ *  - relatives paths are not possible
```
```suggestion
 *  - relative paths are not possible
```

simonsasse commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,334 @@
+ *
+ * Caveat:
+ *  - relatives paths are not possible
+ *  - sandboxed_path is not 100% tight
```
What is a 100% tight path vs. a not 100% tight path ? :D

simonsasse commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,334 @@
+     * \param path must be an absolute path.
+     *
+     * A sandboxed_path initialized with this constructor will
+     * point to path and disallow extension that leave path.
```
maybe "disallow an extension that leaves the path" or "disallow extensions that leave the  path" is a bit clearer.

simonsasse commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,334 @@
+private:
+    void normalize()
+    {
+        auto concated_path   = sandbox_path / std::filesystem::path{*this};
```
```suggestion
        auto concatenated_path   = sandbox_path / std::filesystem::path{*this};
```

simonsasse commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,334 @@
+    void normalize()
+    {
+        auto concated_path   = sandbox_path / std::filesystem::path{*this};
+        auto normalized_path = concated_path.lexically_normal();
```
same here :)

simonsasse commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,334 @@
+    void swap(sandboxed_path &) noexcept = delete; //!< Swap operator is not possible
+
+private:
+    void clear() = delete; //!< Unlikly operator
```
```suggestion
    void clear() = delete; //!< Unlikely operator
```

simonsasse reviewed (changes requested):
Just some typos. Everything else looks good to me :)

SGSSGene commented:
@simonsasse Thanks for your feedback.
I have corrected all the typos. I also clarified a few more than you requested. It would be great if you could take another look.
Also, I changed a "detail" on how the `sandboxed_path` works. Now it has to store absolute paths, otherwise it is way to easy
to break out of the sandbox.

The gcc7 is failing, I have no solution for it yet. I am still thinking about it. But I guess what I have to do is to implement
`weakly_canonical` and `lexically_relative` for gcc7.
Maybe it is possible to have a less strict version of `sandboxed_path` in gcc7, with less good implementations of the missing functions. (or leaving them out all together)

simonsasse commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,344 @@
+private:
+    /*!\brief Normalizes the path.
+     *
+     * Normalization means that is converted to a absolute path and
+     * the path is lexically normalized as described in std::filesystem::path
```
```suggestion
     * Normalization means that the path is converted to an absolute path and
     * lexically normalized as described in std::filesystem::path
```

simonsasse reviewed (changes requested):
@SGSSGene Just a tiny thing. :)
It looks like your test fails not only for gcc7 but also for the other versions now.

SGSSGene commented:
Resolution: don't check for invariant for gcc7

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,416 @@
+    void clear() = delete; //!< Unlikely to be usefull, not implemented.
+};
+
+/** Free sandboxed_path append operator.
+ *  This work the same as std::filesystem::operator/(std::filesystem::path&)
+ *  and additionally checks the invariant.
+ */
+template <typename Rhs>
+sandboxed_path operator/(sandboxed_path lhs, Rhs const& rhs) {
+    lhs /= rhs;
+    return lhs;
+}
```
You should be able to remove this.

std::filesystem::operator/(std::filesystem::path&) works via ADL and ADL also looks for base classes.

SGSSGene commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,416 @@
+    void swap(sandboxed_path &) noexcept = delete; //!< Swap operator is not possible
+
+private:
+    void clear() = delete; //!< Unlikely to be usefull, not implemented.
```
I changed it to public. The issue is that the invariant requires that the current path is an absolute path. calling `clear()` will always violate the invariant.

SGSSGene commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,416 @@
+    }
+
+
+    void swap(sandboxed_path &) noexcept = delete; //!< Swap operator is not possible
```
true, should work :-)

simonsasse reviewed (approved):
Looks good 👍

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,461 @@
+#endif
+    }
+
+    /*!\brief Replaces the path with a new path.
+     * \tparam source_t The type of the source.
+     *
+     * This works the same as std::filesystem::path::assign
+     * and additionally checks the invariant.
+     *
+     * ### Exceptions
+     *
+     * Basic exception guarantee
+     */
+    template <typename source_t>
+    sandboxed_path & assign(source_t const & source)
+    {
+        std::filesystem::path::assign(source);
+        normalise();
+        checkInvariant();
+        return *this;
+    }
```
Maybe you can think of another name than `source` and `source_t`?
It's from the standard, but it would be probably more clear if we used `path_t` and `new_path` or something like this?

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,461 @@
+    }
+
+    /*!\brief Replaces the path with a new path.
+     * \tparam source_t The type of the source.
```
```suggestion
     * \tparam source_t The type of the new path.
```

eseiler replied:
`\param` is missing

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,461 @@
+    void clear() = delete; //!< Not implemented. Invariant requires the path to be an absolute path.
+};
+
+/** Free sandboxed_path append operator.
+ *  This work the same as std::filesystem::operator/(std::filesystem::path&)
+ *  and additionally checks the invariant.
+ */
+template <typename Rhs>
+sandboxed_path operator/(sandboxed_path lhs, Rhs const& rhs) {
+    lhs /= rhs;
+    return lhs;
+}
```
```suggestion
/*!\brief Append a path to a seqan3::test::sanboxed_path.
 * \tparam path_t The type of the path to append.
 * \param lhs The seqan3::test::sandboxed_path.
 * \param rhs The path to append.
 * \relates seqan3::test::sandboxed_path
 *
 *  This work the same as std::filesystem::operator/(std::filesystem::path&)
 *  and additionally checks the invariant.
 */
template <typename path_t>
sandboxed_path operator/(sandboxed_path lhs, path_t const & rhs)
{
    lhs /= rhs;
    return lhs;
}
```

eseiler reviewed (changes requested):
Our maximum line length is 120 (also in documentation), so feel free to reflow where you think it would help.

We do not build (yet?) documentation for `test/include`, so there are no checks for the documentation. If you still want to check (but this is just if you are curious), you could add the file to INPUT and INCLUDE_PATH: https://github.com/seqan/seqan3/blob/dfdccb460da99d1be8610fce98c28824c1f5b554/test/documentation/seqan3_doxygen_cfg.in#L9

SGSSGene commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,461 @@
+        checkInvariant();
+    }
+
+    sandboxed_path() = delete; //!< Deleted.
```
Can't be done. The sandbox_directory path would be "", which is invalid since we need to have an absolute path.

SGSSGene commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,461 @@
+        std::string normalised_path;
+        for (auto const& p : path_parts) {
+            normalised_path += "/" + p;
+        }
```
I don't want a view as a result, but an actuall string. It could be done with std::accumulate. But that seems quit anoying:
```
auto normalised_path = std::accumulate(begin(path_parts), end(path_parts), std::string{""}, [](auto prev, auto const& next)
{
    return prev + '/' + next;
});
```
Yes it is one line shorter, but one of the lines is way longer ;-) Not sure if it is what we want.

SGSSGene commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,461 @@
+        // Now we need to manually collapse any "." and ".."
+        size_t current_pos = path.find("/", 0); // find first "/" character
+        std::vector<std::string> path_parts;
+        while (current_pos < path.size()) {
```
I am not sure how this works. I tried:
```
#include <seqan3/std/ranges>
....
for (auto word : path_string | std::views::spit('/')) {
// unclear what to do with word
}
```
`word` is of type `ranges::detail::split_outer_iterator<ranges::split_view<ranges::ref_view<const std::__cxx11::basic_string<char> >, ranges::single_view<char> >, true>`
The documentation is to spare for me to make anything useful out of it. Any hints?

SGSSGene commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,461 @@
+ * Caveat:
+ *  - relative paths are not possible
+ *  - some functions will leave the sandboxed environment.
+ *    - e.g.: calling relative_path() leaves the environment of sandboxed_path.
```
`relative_path()` takes a `std::filesystem::path` because of slicing it will accept a `sandboxed_path` and returns a `std::filesystem::path`. This returned `std::filesystem::path` is of course outside of the handle of `sandboxed_path`. No gurantees we can make. We could offer our own `relative_path()` function, but it doesn't really make sense, since one of the constraints of `sandboxed_path` is that we operate on absolute paths (to avoid other issues).
All in all it would take a lot more work to get a 100% tight solution while being a drop in replacement for std::filesystem::path and easy to use.

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,562 @@
+    }
+
+    sandboxed_path() = delete;                           //!< Deleted.
+    sandboxed_path(sandboxed_path const&) = default;     //!< Defaulted.
+    sandboxed_path(sandboxed_path&&) noexcept = default; //!< Defaulted.
```
Copy and move assignment missing

SGSSGene replied:
They are directly below the constructors, maybe not obviously but they are covered by the template copy/move assignments.
```
    template <typename path_t>
    sandboxed_path & operator=(path_t const & new_path)
...
```

eseiler replied:
We still add the non-template versions.

SGSSGene replied:
Ah, ok, than I will add them :-)

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,562 @@
+        check_invariant();
+    }
+
+    sandboxed_path() = delete;                           //!< Deleted.
```
Why deleted?
I would only delete the default constructor if constructed object is ill-formed, if it just doesn't make sense to default construct it, I would still leave it.

SGSSGene replied:
It is ill-formed. It would always throw an exception inside the constructor.

eseiler replied:
Where would it throw?

`sandbox_path() = default;` should default construct the `sandbox_directory` and this shouldn't throw.

It will be nonsense since you cannot change the directory, but default-constructability is a nice trait to have.

SGSSGene replied:
I can't do `sandbox_path() = default`. I have to do at least:
```
sandbox_path() {
check_invariant();
}
```
and the `check_invariant()` would through.

I could of course say `sandbox_path() : sandbox_path('/') {}` and construct it to always be the root directory. But that seems rather suprising.

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,562 @@
+     * \param str String that is going to be split
+     * \param delim Delimiter at which point the string should be split.
+     */
+    static std::vector<std::string_view> split_string(std::string const& str, char delim) {
+        std::vector<std::string_view> result{};
+        // If string is empty, nothing to split
+        if (str.empty()) {
+            return result;
+        }
+
+        size_t current_pos = str.find(delim, 0);
+
+        // If string doesn't contain any `delim` return view to the string
+        if (current_pos == std::string::npos) {
+            result.emplace_back(str);
+            return result;
+        }
+
+        while (current_pos < str.size())
+        {
+            auto end_pos = std::min(str.find(delim, current_pos + 1), str.size());
+            auto begin   = str.data() + current_pos + 1;
+            auto length  = end_pos - current_pos - 1;
+            result.emplace_back(begin, length);
+            current_pos = end_pos;
+        }
+        return result;
+    }
```
```suggestion
    static std::vector<std::string_view> split_string(std::string const & str, char const delim)
    {
        std::vector<std::string_view> result{};
        // If string is empty, nothing to split
        if (str.empty())
        {
            return result;
        }

        size_t current_pos = str.find(delim, 0);

        // If string doesn't contain any `delim` return view to the string
        if (current_pos == std::string::npos)
        {
            result.emplace_back(str);
            return result;
        }

        while (current_pos < str.size())
        {
            auto end_pos = std::min(str.find(delim, current_pos + 1), str.size());
            auto begin   = str.data() + current_pos + 1;
            auto length  = end_pos - current_pos - 1;
            result.emplace_back(begin, length);
            current_pos = end_pos;
        }
        return result;
    }
```

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,562 @@
+        std::vector<std::string_view> result{};
+        // If string is empty, nothing to split
+        if (str.empty()) {
+            return result;
```
coverage complains

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,562 @@
+                {
+                    throw std::filesystem::filesystem_error("Path can not be normalised",
+                                                            *this,
+                                                            std::make_error_code(std::errc::invalid_argument));
```
coverage complains

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,562 @@
+        std::filesystem::path::append(first, second);
+        normalise();
+        check_invariant();
+        return *this;
```
coverage complains

eseiler reviewed (changes requested):
💅 
Can you check codecov?

eseiler commented on test/include/seqan3/test/sandboxed_path.hpp:
```diff
@@ -0,0 +1,600 @@
+     *
+     * Basic exception guarantee.
+     */
+    sandboxed_path & operator=(sandboxed_path const & new_path)
+    {
+        std::filesystem::path::operator=(new_path);
+        check_invariant();
+        return *this;
+    }
+
+    /*!\brief Replaces the path with a new path.
+     * \param  new_path The new path.
+     *
+     * This works the same as std::filesystem::path::operator=
+     * and additionally checks the invariant.
+     *
+     * ### Exceptions
+     *
+     * Basic exception guarantee
+     */
+    sandboxed_path & operator=(sandboxed_path && new_path)
+    {
+        std::filesystem::path::operator=(std::move(new_path));
+        check_invariant();
+        return *this;
+    }
```
Wouldn't `=default;` be ok here because the `new_path` must have had checked its invariant at some point?

SGSSGene replied:
This would only be possible if `this->sandbox_directory == new_path.sandbox_directory` was true. But `new_path` can have a path to a different sandbox. The way I designed this class is that `sandbox_directory` is unchangeable, even under a move.

eseiler replied:
Ah, right :)

eseiler reviewed (changes requested):
One last thing :)
